feat!: Async Document Renaming

gavindsouza opened the pull request:
### Problem

Renaming documents can be an expensive affair. The time taken for this operation will increase as your site grows. Depending on the # of links and sizes of tables, this easily exceeds the timeout limits for your Gunicorn or/and NGINX workers. This wouldn't be a problem with ASGI...but that would bring its own host of problems 😆 

### Solution

Document Renames will be pushed to the background worker. This is equally heavy on UX as it is with correctness & sanity. Let's hope we get this right 🤞🏼 

### Changes

* Added whitelisted `Document.rename` API (via 5a3d8f925ee93a393ff5143d372b4e97bdfa4bb2)

Use Document API:

```python
In [1]: doc = frappe.get_doc("Person", "5a188f66c1")

In [2]: doc.rename("5a188f66c2")

In [3]: doc.name
Out[3]: '5a188f66c2'
```

Enqueue Action:

```python
In [1]: doc = frappe.get_doc("Person", "5a188f66c1")

In [2]: doc.queue_action("rename", name="5a188f66c2")
Out[3]: Job('c8d7fa3a-8d58-4388-aad2-2a9dc6153b14', enqueued_at=datetime.datetime(2022, 2, 24, 20, 8, 44, 940008))
```

* Refactored `doc.queue_action`, `execute_action` (via 040e01eb59b90b7c2b0ebf103324a83065306c08)
* Performance enhancements
* Converted raw SQLs in rename_doc module to QB notation

### Goal

The goal of this PR is to manage asynchronous renames via the `update_document_title` API which is triggered via the toolbar options to rename document / update title.

<!-- no-docs -->

gavindsouza committed:
perf: Pass document through rename_doc API to bypass extra get_doc

gavindsouza committed:
feat: Document.rename API

Transform current document object using the rename_doc API. The design
of the API should allow for easy action queueing.

Defined as `rename(self, name: str, merge: bool = False, force: bool = False)`

Usage:

    In [1]: doc = frappe.get_doc("Person", "5a188f66c1")

    In [2]: doc.rename("5a188f66c2")

    In [3]: doc.name
    Out[3]: '5a188f66c2'

gavindsouza committed:
refactor: doc.enqueue_action & execute_action

* Make args "internal" to avoid name collision. Added __ since these
  args would be utilized by the internal API alone
* perf(EAFP): Try taking a lock and then figure out if there's an error
  instead of the double computation
* Notify document update on completion of Document Action

gavindsouza committed:
test: Added test_doc_rename

gavindsouza committed:
fix: Use rename doc method wrapper

gavindsouza committed:
test(fix): Add property setter for ToDo to be made rename-able

gavindsouza committed:
fix: Print to STDOUT to help debug via is_scheduler_inactive

gavindsouza committed:
feat: Rename Document via Background Job

gavindsouza commented:
Thinking of adding an argument that specifies whether to rename the document in the background job or in the current process. The current one feels too magical 🤔

gavindsouza committed:
test(fix): Add cleanup for test_doc_rename_method

gavindsouza committed:
fix: enqueue option for update_document_title API

Pass enqueue=True to enqueue rename document operation. Desk will be
using this from now on ;)

gavindsouza committed:
fix: Remove taken file locks at end of background job

gavindsouza committed:
Merge branch 'develop' into bg-rename_doc

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into bg-rename_doc

gavindsouza committed:
fix: Remove lock record if exists

gavindsouza committed:
fix: Setup queued renaming listener based on API response

ankush commented on frappe/public/js/frappe/form/toolbar.js:
```diff
@@ -103,10 +102,29 @@ frappe.ui.form.Toolbar = class Toolbar {
+						}
+					});
+					frappe.show_alert(
+						__('Document renaming from {0} to {1} has been queued', [docname.bold(), new_name.bold()])
```
```suggestion
						__('Document renaming from {0} to {1} has been queued, please check after some time.', [docname.bold(), new_name.bold()])
```

"queued" is bit technical, can we add a bit more info?

---

Minor UX issues around locking:

Immediately re-renaming is allowed even if the previous renaming isn't finished. (poor locking implementation I guess?) Perhaps we should immediately make the form read-only or `frm.disable_save` on queued docs with some message for why it's read-only. I think we need to utilize `HTTP 202` for this 😄 


PS:  I just added 100s sleep to simulate long renames.

ankush replied:
ERPNext has some longish complicated rename tests if reqd.

ankush replied:
Another thing to consider. Some renames have [pre-rename validations](https://github.com/ankush/erpnext/blob/90e4b0bdd9746590bd44a3d622168a2f98bb5129/erpnext/stock/doctype/item/item.py#L395-L402), right now it seems the messages are just posted as comments on doc.

The problem is we can't just run before_rename too since that can also have some non-validation actions => partial commits 😐

gavindsouza replied:
> "queued" is bit technical, can we add a bit more info?

We could; but I'm not sure about this. When the renaming is completed successfully, the user is re-directed to the document. I'm not sure about that either since we've not really standardized these things yet 🤔. Hoping that enqueuing actions becomes first-class soon in Frappe 🥲 

> Immediately re-renaming is allowed even if the previous renaming isn't finished. (poor locking implementation I guess?) Perhaps we should immediately make the form read-only or `frm.disable_save` on queued docs with some message for why it's read-only. I think we need to utilize `HTTP 202` for this 😄

Couldn't replicate this. Did you restart your bench before trying this? I tried the same and got a error saying that the Document is locked.

> Another thing to consider. Some renames have [pre-rename validations](https://github.com/ankush/erpnext/blob/90e4b0bdd9746590bd44a3d622168a2f98bb5129/erpnext/stock/doctype/item/item.py#L395-L402), right now it seems the messages are just posted as comments on doc.
> 
> The problem is we can't just run before_rename too since that can also have some non-validation actions => partial commits 😐

Moved validations out of the queued job, made sure they just run once whether in or out. Thought of throwing an exception if a commit was triggered in the `before_rename` but I feel I may have already gone beyond the scope of this PR.

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into bg-rename_doc

gavindsouza committed:
fix: validate before enqueuing rename_doc

* refactor validate_rename
* don't run before_rename hooks twice
* validate_rename kwarg in doc.rename

gavindsouza committed:
fix: Validate title updates via update_document_title API

gavindsouza committed:
fix: Check if name was changed in client

gavindsouza committed:
chore: Add docstring for rename_doc

gavindsouza committed:
fix(rename_doc): Use sbool instead of cint

cint("false") returns True which is what is sent by frappe dialog. This
may be required to be fixed in the client alone but making this change
to make the API more "robust" as this has been working in this
particular way for far too long now :')

gavindsouza committed:
refactor(rename_doc): Use QB notation inplace of raw SQLs

* Converted ~22 queries from raw SQL to use frappe.qb notation
* Made queries DRY-er

gavindsouza committed:
fix(qb): Make Table importable

ankush commented on frappe/model/document.py:
```diff
@@ -938,6 +938,14 @@ def _cancel(self):
 
+	@whitelist.__func__
+	def _rename(self, name: str, merge: bool = False, force: bool = False, validate_rename: bool = True):
+		"""Cancel the document. Sets `docstatus` = 2, then saves.
```
```suggestion
```

incorrect docstring.

ankush commented on frappe/model/rename_doc.py:
```diff
@@ -47,11 +63,33 @@ def update_document_title(
+			current_name = doc.name
+
+			# before_name hook may have DocType specific validations or transformations
+			transformed_name = doc.run_method("before_rename", current_name, updated_name, merge)
```
potential of non-atomic updates here:

Example: https://github.com/frappe/erpnext/blob/707f624aabb6721d6a1817f75dbc1c17f8c5cf81/erpnext/stock/doctype/item/item.py#L395-L403

ankush replied:
Maybe we should think of UX of validations later and ensure atomicity first 🥲 

Or new hook `validate_hook` that just does validation and no DB writes.

gavindsouza replied:
😭

gavindsouza committed:
fix: rollback to savepoint to avoid partial commits

It's better to keep the validations in and out AND separate...this is a
humble attempt for the same :)

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into bg-rename_doc

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into bg-rename_doc

gavindsouza committed:
fix: Generate hash of length 8 for save point

gavindsouza committed:
refactor: rename_document_title

* Remove extra reload_doc that caused extra onload
* Refactor variable naming, DRY-er
* Simplify flow

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into bg-rename_doc
